src/api/httpClient.ts

```diff
@@ -0,0 +1,371 @@
+                        headers: {
+                            // eslint-disable-next-line @typescript-eslint/naming-convention
+                            Token: this.poliNetworkToken?.refresh_token,
+                        },
+                        retryType: RetryType.RETRY_N_TIMES,
+                        maxRetries: 5,
+                    }
+                )
+            if (typeof response.data.access_token === "string") {
+                console.log("Refreshed polinetwork token")
+
+                this.poliNetworkToken = response.data
+
+                // save back the tokens
+                const tokens: Tokens = {
+                    poliNetworkToken: this.poliNetworkToken,
+                    polimiToken: this.polimiToken,
+                }
+                await AsyncStorage.setItem("api:tokens", JSON.stringify(tokens))
+
+                return true
+            } else {
+                console.warn("Invalid response refreshing polinetwork token")
+                console.warn(response.status)
+                console.warn(response.data)
+                return false
+            }
+        } catch (e) {
+            console.warn("Error refreshing polinetwork token")
+            console.warn(e)
+            return false
+        }
+    }
+    /**
+     * load tokens from storage on boot, if they aren't present, do nothing
+     */
+    async loadTokens() {
+        const tokens = await AsyncStorage.getItem("api:tokens")
+        if (tokens) {
+            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
+            const parsedTokens: Tokens = JSON.parse(tokens)
+            console.log("Loaded tokens from local storage")
+            this.polimiToken = parsedTokens.polimiToken
+            this.poliNetworkToken = parsedTokens.poliNetworkToken
+            this.emit("login")
+            this.emit("login_event", true)
+        } else {
+            console.log("No tokens found in local storage")
+        }
+    }
+    /**
+     * set the tokens and save them to storage
+     * @param tokens both the polinetwork and polimi tokens
+     */
+    async setTokens(tokens: Tokens) {
+        this.polimiToken = tokens.polimiToken
+        this.poliNetworkToken = tokens.poliNetworkToken
+
+        this.emit("login")
+        this.emit("login_event", true)
+
+        // save the tokens in local storage
+        await AsyncStorage.setItem("api:tokens", JSON.stringify(tokens))
+        console.log("Saved tokens in local storage")
+    }
+    /**
+     * remove the tokens from storage, essentially log out
+     */
+    async destroyTokens() {
+        console.log("Destroying tokens, logging out")
+
+        this.polimiToken = undefined
+        this.poliNetworkToken = undefined
+
+        this.emit("logout")
+        this.emit("login_event", false)
+
+        // remove the tokens from local storage
+        await AsyncStorage.removeItem("api:tokens")
+    }
+}
+
+export const client = HttpClient.getInstance()
```

Comment:
sai cosa, così mi piace, ma dato che a questo punto il client http diventa una cosa che deve essere toccata solo internamente dall'API, secondo me ha senso fare embracing della natura di singleton della classe, so che non cambia un cazzo però mi sembra più bello:

esporta la classe e lascia che siano le singole chiamate a prendere l'istanza del client (non è il nostro caso, ma concettualmente l'istanza potrebbe non essere la stessa per qualche ragione) chiamando direttamente `getInstance()` tipo così
```ts
import { HttpClient } from "api/HttpClient"

...

const client = HttpClient.getInstance()
const response = await client.polimiInstance.get("/path/")
```
direi che si può rinominare il file usando di nuovo PascalCase, dato che il focus principale è quello di esporre la classe in se (se vogliamo essere più puritani possiamo anche spostare le altre interface in un altro file ma le vedo come abbastanza rilevanti per essere lasciate qui)

```suggestion
```

dimmi cosa ne pensi

Reply:
Si va bene, mi sembra più pulito e come hai detto sfruttiamo il fatto che HttpClient è un singleton, se no tanto valeva fare un oggetto normale ahhaha

Reply:
Ho dovuto cambiare l'import del client in Login.tsx Settings.tsx App.tsx e loadTokens.ts a favore di HttpClient.getInstance() poichè veniva utilizzato per settare e distruggere token etc.